lib/services/local/plugin_data.go

```diff
@@ -0,0 +1,219 @@
+	return trace.Wrap(s.upsertPluginData(ctx, params, parentExpiresAt))
+}
+
+// updatePluginData updates or creates PluginData using CAS
+func (s *PluginDataService) upsertPluginData(ctx context.Context, params types.PluginDataUpdateParams, parentExpiresAt *time.Time) error {
+	retryPeriod := retryPeriodMs * time.Millisecond
+	retry, err := utils.NewLinear(utils.LinearConfig{
+		Step: retryPeriod / 7,
+		Max:  retryPeriod,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Update is attempted multiple times in the event of concurrent writes.
+	for i := 0; i < maxCmpAttempts; i++ {
+		var create bool
+		var data types.PluginData
+		item, err := s.Get(ctx, pluginDataKey(params.Kind, params.Resource))
+		if err == nil {
+			data, err = itemToPluginData(*item)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			create = false
+		} else {
+			if !trace.IsNotFound(err) {
+				return trace.Wrap(err)
+			}
+			// In order to prevent orphaned plugin data, we automatically
+			// configure new instances to expire shortly after the resource
+			// to which they are associated.  This discrepency in expiry gives
+			// plugins the ability to use stored data when handling an expiry
+			// (OpDelete) event.
+			if parentExpiresAt == nil {
+				return trace.BadParameter("missing parent resource %q", params.Kind)
+			}
+			exp := *parentExpiresAt
+
+			data, err = types.NewPluginData(params.Resource, params.Kind)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			// A resource can have empty expiration time by default.
+			// In this case, we should leave PluginData expiration time empty as well.
+			// TODO: PluginData must be destroyed with parent resource. This needs to be addressed in future.
+			if !exp.IsZero() {
+				data.SetExpiry(exp.Add(time.Hour))
+			}
+			create = true
+		}
+		if err := data.Update(params); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := data.CheckAndSetDefaults(); err != nil {
+			return trace.Wrap(err)
+		}
+		newItem, err := itemFromPluginData(data)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if create {
+			if _, err := s.Create(ctx, newItem); err != nil {
+				if trace.IsAlreadyExists(err) {
+					select {
+					case <-retry.After():
+						retry.Inc()
+						continue
+					case <-ctx.Done():
+						return trace.Wrap(ctx.Err())
+					}
+				}
+				return trace.Wrap(err)
+			}
+		} else {
+			if _, err := s.CompareAndSwap(ctx, *item, newItem); err != nil {
+				if trace.IsCompareFailed(err) {
+					select {
+					case <-retry.After():
+						retry.Inc()
+						continue
+					case <-ctx.Done():
+						return trace.Wrap(ctx.Err())
+					}
+				}
+				return trace.Wrap(err)
+			}
+		}
+		return nil
+	}
+	return trace.CompareFailed("too many concurrent writes to plugin data %s", params.Resource)
+}
+
+// pluginDataKey constructs PluginData entry backend key
+func pluginDataKey(kind string, name string) []byte {
+	return backend.Key(pluginDataPrefix, kind, name, paramsPrefix)
+}
```

Comment:
> 2. AccessRequests have expiration time. PluginData TTL is explicitly set to parent AccessRequest.Expires + 1h on create. This ensures that there would be no orphaned PluginData. Normally, user does not have any TTL. So, PluginData is created with no expiration by default. That might become a problem. However, I am not sure that this is the scope of this PR.

The way this has been handled in the past with other resources which could be orphaned once a user is deleted (such as user recovery tokens) is to make the resource key prefix have the parent resource key as a prefix. Ex: `/web/users/{user_id}`, `/web/users/{user_id}/recoverycodes/{code}`. This way, when the user is deleted, any resource with `web/user/{user_id}` as a key prefix is deleted as well.

```suggestion
// pluginDataKey constructs PluginData entry backend key
func pluginDataKey(kind string, name string) []byte {
	if kind == types.KindUser {
		return backend.Key(webPrefix, usersPrefix, name, pluginDataPrefix, paramsPrefix)
	}
	return backend.Key(pluginDataPrefix, kind, name, paramsPrefix)
}
```

Reply:
I realised that the case I described above: when PluginData needs to be around for some time after a parent object gets destroyed is not be applicable to users. So, I replaced it here like you proposed. ^